design/multiple-listeners-design.md

```diff
@@ -0,0 +1,147 @@
+
+## Non Goals
+- Support for HTTPProxy/Ingress attaching to additional Listeners.
+- Support for new route types, e.g. TCPRoute.
+
+## High-Level Design
+Broadly speaking, Contour can operate in one of two modes, which we will call **Gateway mode** and **non-Gateway mode**.
+In Gateway mode, Contour is configured to process a Gateway API Gateway, and gets its Listener configuration from there.
+Non-Gateway mode is the traditional mode of operation, where Listener configuration comes from CLI flags, a config file, and/or a ContourConfiguration resource, plus the Envoy service YAML, and HTTPProxy and/or Ingress is used for routing configuration.
+
+For non-Gateway mode, this design should be fully backwards-compatible.
+There should be no changes observed by a user of Contour in non-Gateway mode.
+
+For Gateway mode, the following changes are needed:
+
+- The Gateway provisioner must configure ports on the Envoy service and daemonset/deployment corresponding to all of the Gateway Listener ports. The mapping between service ports and container ports will be described in the Detailed Design section below.
+- Contour's directed acyclic graph (DAG) processing changes, to first add Listeners, and then add virtual hosts directly to the appropriate Listener(s).
+- Each DAG Listener results in an Envoy Listener being configured. The naming scheme will be described in the Detailed Design section below. 
+- Envoy RouteConfigurations will be generated for each Envoy Listener. The naming scheme will be described in the Detailed Design section below. 
+
+For Gateway mode, we must also consider what happens when HTTPProxy or Ingress is used for routing configuration.
+The details of this use case are described in the Compatibility section below.
+
+## Detailed Design
+
+### internal/gatewayapi
+The `ValidateListeners` function logic must change, to allow any number of unique ports, while still validating protocols and hostnames for each port.
+Instead of returning a single insecure and secure port number, it will return a slice of ports.
+
+### internal/provisioner
+The Gateway controller will use the results of the `ValidateListeners` function (above) to add all relevant ports to `contourModel.Spec.NetworkPublishing.Envoy.Ports`.
+These ports will be populated on the Envoy service and deployment/daemonset.
+
+Currently, the single insecure port is always mapped to port 8080 inside the Envoy container, and the single secure port is always mapped to 8443.
+Supporting any number of Listeners requires a more flexible port mapping scheme, since the requested Gateway Listener ports could be privileged ports (e.g. 1-1023), non-privileged ports (e.g. 8080, 8443), and/or node ports (e.g. 32080).
+The main constraint is that, since Envoy runs as non-root, it cannot bind to privileged ports within the container.
+As such, any requested privileged Gateway Listener port must be mapped to a non-privileged port inside the Envoy container.
+The scheme proposed is as follows:
```

Comment:
though conflicts here might be an edge case, we could try to prevent them entirely by setting container ports to increment with the index of the unique Gateway Listener ports, e.g. if they are unique `spec.listeners[0]` gets port 1024 in the container, `spec.listeners[1]` gets 1025, and so on.

this might be a little harder to keep straight in terms of debugging etc., but would reduce some error handling complication etc. since [you can only have up to 64 Listeners at the moment](https://github.com/kubernetes-sigs/gateway-api/blob/e1b21a5ba165c19fc2be50a3599b607e4834cf1a/apis/v1beta1/gateway_types.go#L113) so running out of ports is unlikely

Reply:
Hmm, worth considering, though yeah I do worry about the associated cost if Listeners come and go, and that triggering port re-mapping. The nice thing about a deterministic mapping based on port number is that they will never change.

Reply:
The other tricky thing is that multiple Gateway Listeners may correspond to a single port / Envoy Listener, so you'd have to figure out how to handle that case. It's not always 1:1.

Reply:
Another option here is to add a static value to all Gateway Listener ports that avoids privileged ports plus the two static Listener ports, so e.g. +9001. I think this is what EG does.

Reply:
yeah that makes a lot of sense too, not really a chance to run out of ports since a single Gateway has a pretty low limit

Reply:
regardless of what we pick, we're aware of the fact we need to map ports and avoid conflicts etc. 👍🏽